src/inc/quic_tls.h

```diff
@@ -198,6 +198,12 @@ typedef struct CXPLAT_TLS_CONFIG {
     //
     QUIC_TLS_SECRETS* TlsSecrets;
 
+    //
+    // Storage for TLS traffic secrets when the connection has the parameter set
+    // to allow encryption offloading
+    //
+    QUIC_TLS_OFFLOAD_SECRETS* TlsOffloadSecrets;
```

Comment:
Shouldn't this already be stored on the keys themselves?

Reply:
You mean inside QUIC_PACKET_KEY defined as WriteKeys/ReadKeys?

Reply:
We discussed over IM that you make a new function that takes a packet key and then generates the key info you need to offload:
```c
_IRQL_requires_max_(DISPATCH_LEVEL)
QUIC_STATUS
QuicPacketKeyCreateOffload(
    _In_ QUIC_PACKET_KEY *Key,
    _In_ const QUIC_HKDF_LABELS* HkdfLabels,
    _In_z_ const char* const SecretName,
    _Out_ CXPLAT_QEO_CONNECTION* Qeo
    )
```
With this model, you don't need to save anything else.

Reply:
As I originally wrote as in https://github.com/microsoft/msquic/blob/b1a6ecf5615f01c0dbd588af0802c101060343bc/src/platform/tls_schannel.c#L3247-L3334 
It requires at least `TlsContext->Workspace.OutSecBuffers[i]` to regenerate keys and it doesn't exist on openssl implementation. So I newly add "cache" which stores keys when it is generated during handshake.
Is there any way to retrieve keys from `QUIC_PACKET_KEY`?

Reply:
Temporally generated keys in QuicPacketKeyDerive is used to create key objects and immediately discarded. As far as I checked there is no invert operation to retrieve raw keys from the key objects.